torch/nn/utils/parametrizations.py

```diff
@@ -84,6 +84,11 @@ def _power_method(self, weight_mat: torch.Tensor, n_power_iterations: int) -> No
 
         # Precondition
         assert weight_mat.ndim > 1
+
+        if n_power_iterations > 0:
+            # See above for why we need this clone
+            self._u = self._u.clone(memory_format=torch.contiguous_format)
```

Comment:
I am not sure to understand what's the point of doing the normalize inplace below if we clone here? Why not just make them out of place?

Also the comment above seem to mention that this could be a problem with DP, do we have a test for that? and a test for DP + two forwards?

Reply:
I don't know what the perf implications are, but if we do multiple iterations of power_iteration, we only do a single out-of-place op if clone.
But since most people have n_power_iterations=1, I can see why it's better to just do out-of-place.
Or I suppose if we really wanted to optimize both cases, we could just add a check: if n_power_iterations > 1, clone. Otherwise, just do it out-of-place.

Yeah from reading the comment in the again, seems like we have to do the modification in-place for DDP to work? I think another alternative is to replace the self.u/self.v with cloned versions AFTER using them in operations that would save them for backward.

It looks like there is a dataparallel test in `test_new_spectral_norm`, we can try to verify this by seeing if CI fails with this current PR.

Reply:
After a quick search of `torch.nn.utils.spectral_norm(` on github, out of the 30+ occurrences I have seen, none of them uses `n_power_iterations != 1`. I believe that it wouldn't be too bad to remove this optimisation if it's causing so much trouble.

All in all, I cannot think of any application where `n_power_iterations > 2` would be necessary, and making an extra copy in the case  `n == 2` is not too bad.

Reply:
I would still like to know if this causes desync in DP or not?

Reply:
Looks like this test failure https://app.circleci.com/pipelines/github/pytorch/pytorch/356979/workflows/9e4b4b08-933c-430a-844d-da3c08d10600/jobs/15050920 confirms that we do need to modify u and v in-place for DP to work.

Looks like there is also test that is supposed to test for double forward (and also runs with DP) https://github.com/pytorch/pytorch/blob/9776e1ff2f2658b25c37beaf3f4aea3c8b8bd174/test/test_nn.py#L4210, but it uses .grad underneath (via gradcheck) and computes grad wrt to input only (and not the parameters). The error only reveals itself when we compute grad wrt the actual parameter.